### ISSUE TITLE
Use of syncOnStopOrFailure in JavaScript callback should not result in compilation errors when compiled with -Ydelambdafy:inline

### ISSUE THREAD

**Issue by dneades** [monix/monix#515]
The macro used by `Ack.syncOnStopOrFailure` (`Macros.syncOnStopOrFailure`) fails when the following file is compiled with the `scalac`  `-Ydelambdafy:inline` option:

_src/main/scala/com/araxis/example/BrowserEvents.scala_:
```
package com.araxis.example

import monix.execution.Scheduler.Implicits.global
import monix.execution.cancelables.SingleAssignCancelable
import monix.execution.{Ack, Cancelable}
import monix.reactive.Observable
import monix.reactive.OverflowStrategy.Unbounded
import org.scalajs.dom.{Event, EventTarget}

import scala.scalajs.js

object BrowserEvents {

  def eventListener(target: EventTarget, event: String): Observable[Event] =
    Observable.create(Unbounded) { subscriber =>
      val c = SingleAssignCancelable()
      // Forced conversion, otherwise canceling will not work!
      val f : js.Function1[Event, Ack] = (e: Event) => subscriber.onNext(e).syncOnStopOrFailure(_ => c.cancel())

      target.addEventListener[Event](event, f)
      c := Cancelable(() => target.removeEventListener[Event](event, f))
    }
}
```

_build.sbt_:
```
enablePlugins(ScalaJSPlugin)

scalaVersion := "2.12.4"

scalacOptions ++= Seq(
  "-Ydelambdafy:inline", // causes Monix JavaScript event listener code not to compile
)

libraryDependencies ++= Seq(
  "org.scala-js" %%% "scalajs-dom" % "0.9.4",
  "io.monix" %%% "monix" % "3.0.0-M3"
)
```

_project/plugins.sbt_:
```
addSbtPlugin("org.scala-js" % "sbt-scalajs" % "0.6.21")
```

_project/build.properties_:
```
sbt.version=1.1.0
```

The compiler output using the above code and configuration is:
```
[info] Compiling 1 Scala source to C:\Hg\Test\target\scala-2.12\classes ...
[error]
[error]   Could not generate raw function for JS function: Found a non-private field value c$1 in final <synthetic> class anonfun$1 extends scala.scalajs.js.Function with scala.scalajs.js.Function1 {
[error]   final def apply(arg1: org.scalajs.dom.raw.Event): monix.execution.Ack = {
[error]     val source$macro$1: monix.execution.Ack = anonfun$1.this.subscriber$1.onNext(arg1);
[error]     if (source$macro$1.eq(monix.execution.Ack$Stop))
[error]       try {
[error]         anonfun$1.this.c$1.cancel()
[error]       } catch {
[error]         case (ex @ (_: Throwable)) => if (monix.execution.misc.NonFatal.apply(ex))
[error]           monix.execution.Scheduler.Implicits().global().reportFailure(ex)
[error]         else
[error]           throw ex
[error]       }
[error]     else
[error]       if ((source$macro$1: scala.concurrent.Future).!=(monix.execution.Ack$Continue))
[error]         source$macro$1.onComplete({
[error]           (new <$anon: Function1>(anonfun$1.this): Function1)
[error]         }, monix.execution.Scheduler.Implicits().global())
[error]       else
[error]         ();
[error]     source$macro$1
[error]   };
[error]   final <bridge> <artifact> def apply(arg1: Object): Object = anonfun$1.this.apply(arg1.$asInstanceOf[org.scalajs.dom.raw.Event]());
[error]   <synthetic> <paramaccessor> val c$1: monix.execution.cancelables.SingleAssignCancelable = _;
[error]   <synthetic> <paramaccessor> private[this] val subscriber$1: monix.reactive.observers.Subscriber$Sync = _;
[error]   def <init>($outer: <$anon: Function1>, c$1: monix.execution.cancelables.SingleAssignCancelable, subscriber$1: monix.reactive.observers.Subscriber$Sync): <$anon: scala.scalajs.js.Function with scala.scalajs.js.Function1> = {
[error]     anonfun$1.this.c$1 = c$1;
[error]     anonfun$1.this.subscriber$1 = subscriber$1;
[error]     anonfun$1.super.<init>(scala.collection.immutable.Nil);
[error]     ()
[error]   }
[error] }
[error]      while compiling: C:\Hg\Test\src\main\scala\com\araxis\example\BrowserEvents.scala
[error]         during phase: jscode
[error]      library version: version 2.12.4
[error]     compiler version: version 2.12.4
[error]   reconstructed args: -Xplugin:C:\Users\djn\.ivy2\cache\org.scala-js\scalajs-compiler_2.12.4\jars\scalajs-compiler_2.12.4-0.6.21.jar -bootclasspath C:\Users\djn\.sbt\boot\scala-2.12.4\lib\scala-library.jar -Ydelambdafy:inline -classpath C:\Hg\Test\target\scala-2.12\classes;C:\Users\djn\.ivy2\cache\org.scala-js\scalajs-library_2.12\jars\scalajs-library_2.12-0.6.21.jar;C:\Users\djn\.ivy2\cache\org.scala-js\scalajs-dom_sjs0.6_2.12\jars\scalajs-dom_sjs0.6_2.12-0.9.4.jar;C:\Users\djn\.ivy2\cache\io.monix\monix_sjs0.6_2.12\jars\monix_sjs0.6_2.12-3.0.0-M3.jar;C:\Users\djn\.ivy2\cache\io.monix\monix-execution_sjs0.6_2.12\jars\monix-execution_sjs0.6_2.12-3.0.0-M3.jar;C:\Users\djn\.ivy2\cache\org.typelevel\cats-core_sjs0.6_2.12\jars\cats-core_sjs0.6_2.12-1.0.1.jar;C:\Users\djn\.ivy2\cache\org.typelevel\cats-macros_sjs0.6_2.12\jars\cats-macros_sjs0.6_2.12-1.0.1.jar;C:\Users\djn\.ivy2\cache\org.typelevel\machinist_sjs0.6_2.12\jars\machinist_sjs0.6_2.12-0.6.2.jar;C:\Users\djn\.sbt\boot\scala-2.12.4\lib\scala-reflect.jar;C:\Users\djn\.ivy2\cache\org.typelevel\cats-kernel_sjs0.6_2.12\jars\cats-kernel_sjs0.6_2.12-1.0.1.jar;C:\Users\djn\.ivy2\cache\io.monix\monix-eval_sjs0.6_2.12\jars\monix-eval_sjs0.6_2.12-3.0.0-M3.jar;C:\Users\djn\.ivy2\cache\org.typelevel\cats-effect_sjs0.6_2.12\jars\cats-effect_sjs0.6_2.12-0.8.jar;C:\Users\djn\.ivy2\cache\io.monix\monix-tail_sjs0.6_2.12\jars\monix-tail_sjs0.6_2.12-3.0.0-M3.jar;C:\Users\djn\.ivy2\cache\io.monix\monix-reactive_sjs0.6_2.12\jars\monix-reactive_sjs0.6_2.12-3.0.0-M3.jar
[error]
[error]   last tree to typer: TypeTree(trait Subscriber$Sync)
[error]        tree position: line 15 of C:\Hg\Test\src\main\scala\com\araxis\example\BrowserEvents.scala
[error]             tree tpe: monix.reactive.observers.Subscriber$Sync
[error]               symbol: abstract trait Subscriber$Sync in package observers
[error]    symbol definition: abstract trait Subscriber$Sync extends Subscriber with Observer$Sync (a ClassSymbol)
[error]       symbol package: monix.reactive.observers
[error]        symbol owners: trait Subscriber$Sync
[error]            call site: <$anon: Function0> in package example in package example
[error]
[error] == Source file context for tree position ==
[error]
[error]     12 object BrowserEvents {
[error]     13
[error]     14   def eventListener(target: EventTarget, event: String): Observable[Event] =
[error]     15     Observable.create(Unbounded) { subscriber =>
[error]     16       val c = SingleAssignCancelable()
[error]     17       // Forced conversion, otherwise canceling will not work!
[error]     18       val f : js.Function1[Event, Ack] = (e: Event) => subscriber.onNext(e).syncOnStopOrFailure(_ => c.cancel())

<truncated for brevity>
```
The above example compiles without error if the `-Ydelambdafy:inline` option is removed.

Finally, note that I am not sure how useful it is to be able to compile with `-Ydelambdafy:inline`, so the practical import of this problem is possibly limited. However, @alexandru requested that I file an issue for this.

**Comment by alexandru** [comment]
@dneades this will be fixed by #517 by turning those macros to plain functions, which should make this problem irrelevant.

I'm trying to keep the issue tracker clean, so I'm going to close this issue, but if you want to add anything, feel free to re-open it or to comment either here or on #517.

**Comment by dneades** [comment]
Thank you very much, @alexandru, that sounds great.
